Support for OpenMM

orionarcher opened the pull request:
## Summary

Include a summary of major changes in bullet points:

* setup functions for OpenMM simulations
* functions to anneal and equilibrate pressure in OpenMM simulations
* includes a fix to packmol also present in PR #3

## Additional dependencies introduced (if any)

openmm: this is code for OpenMM IO

openff: needed to parameterize Sage Force Field

parmed: used to create suitably sized PDB files. Later may be used to add more force fields.

openbabel + rdkit are used, but they are already used in pymatgen.

## TODO (if any)

- [x] better support for water models, at least allow the user to specify TIP3 or TIP4.

- [x] automatic partial charge assignment. It should be possible to include RESP partial charges in the parameterization.

- [x] allow partial charge scaling as a setting in the generator

## Checklist

- [x] Code is in the [standard Python style](https://www.python.org/dev/peps/pep-0008/). The easiest way to handle this
      is to run the following in the **correct sequence** on your local machine. Start with running
      [black](https://black.readthedocs.io/en/stable/index.html) on your new code. This will automatically reformat
      your code to PEP8 conventions and removes most issues. Then run 
      [pycodestyle](https://pycodestyle.readthedocs.io/en/latest/), followed by 
      [flake8](http://flake8.pycqa.org/en/latest/).
- [x] Docstrings have been added in the [Google docstring format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html).
      Run [pydocstyle](http://www.pydocstyle.org/en/2.1.1/index.html) on your code.
- [x] Type annotations are **highly** encouraged. Run [mypy](http://mypy-lang.org/) to type check your code.
- [x] Tests have been added for any new functionality or bug fixes.
- [ ] All linting and tests pass.

orionarcher committed:
initial files for Pymatgen OpenMM IO

orionarcher committed:
added several convenience functions for setting up OpenMM simulations

orionarcher committed:
rename files

orionarcher committed:
added an equilibrate pressure and anneal simulation stages

orionarcher committed:
added code to return a simulation object from simulation generator

orionarcher committed:
improved documentation for the setup classes

orionarcher committed:
added testing files

orionarcher committed:
skeleton for testing setup and simulations

orionarcher committed:
basic type checking testing

orionarcher committed:
fixed an issue where quotation marks inside strings were causing packmol to not recognize file names

orionarcher committed:
fixing broken tests

orionarcher committed:
small formatting changes to setup

orionarcher committed:
bug fixes and finished basic testing for openmm setup

orionarcher committed:
bug fixes and testing for the anneal simulation runner

orionarcher committed:
finished bug fixes on simulation code and finished basic testing

orionarcher committed:
noted bug to be fixed later

rkingsbury commented:
Thanks @orioncohen ! I won't be able to look at this in much detail until early next week but I'm excited to see how your OpenMM support is shaping up.

orionarcher commented:
Sounds good!

I believe that the OpenMM IO can fit into the new IO abstraction pretty cleanly by serializing all of the OpenMM objects needed to create a simulation

The InputSetGenerator will be instantiated by user defined settings and create a Topology, System, Integrator, and Simulation. It will then serialize all of those objects into XML and store them. When `get_input_set` is called, it will create an `InputSet` that has the XML files as attributes. This process matches the rest of Pymatgen IO and is fit for HTMD.

The InputSet can then have a `get_simulation` method that instantiates a simulation and returns it. This way, no OpenMM objects are ever stored and the InputSetGenerator and InputSet remain MSONable.

orionarcher committed:
started adapting the openmm IO to fit the new IO abstraction

orionarcher committed:
corrected format of InputFiles and updated the OpenMM input set

orionarcher committed:
added get_simulation and validate methods to the OpenMMSet

orionarcher commented:
What should determine if a parameter goes in `InputGenerator.__init__` or `InputGenerator.get_input_set`? 

The documentation seems to indicate that only the structure and other critical parameters should go in `get_input_set`, but I can imagine many situations where that would be inconvenient. For example, if I wanted to repeat a simulation at four different temperatures. I'm tempted to put all simulations settings in both `__init__` and `get_input_set` for maximum flexibility, but I'm curious to hear more about the intended design.

orionarcher committed:
small bug fix to setup

orionarcher committed:
update signature of OpenMMGenerator and begin migrating setup methods

orionarcher committed:
added setup methods to OpenMMGenerator class

orionarcher committed:
move from 'generate' for method names to 'get'

orionarcher committed:
finish basic system parameterization for Sage force field

orionarcher committed:
add a reminder to insert partial charge handling

orionarcher committed:
added comments to document development plan

rkingsbury commented on pymatgen/io/openmm/inputs.py:
```diff
@@ -0,0 +1,301 @@
+# noinspection PyMethodOverriding
+class OpenMMGenerator(InputGenerator):
+
+    # TODO: what determines if a setting goes in the __init__ or get_input_set?
```
@orioncohen I would say anything that is _system specific_ goes to `get_input_set` (so typically, just your coordinates) whereas all other settings e.g. temperature, run time, integrator, etc. are part of the generator. I think of the generator as the "recipe" for a particular calculation.

The distinction for MD is admittedly murkier than it is for other types of calculations. For example. I could be convinced that basic things such as step size and run time ought to go into `get_input_set` since they don't have any bearing on the actual content of the calculation (forcefield, etc.). What do you think?

rkingsbury replied:
> 
> 
> What should determine if a parameter goes in `InputGenerator.__init__` or `InputGenerator.get_input_set`?
> 
> The documentation seems to indicate that only the structure and other critical parameters should go in `get_input_set`, but I can imagine many situations where that would be inconvenient. For example, if I wanted to repeat a simulation at four different temperatures. I'm tempted to put all simulations settings in both `__init__` and `get_input_set` for maximum flexibility, but I'm curious to hear more about the intended design.

Personally I am reluctant to duplicate settings in both `__init__` and `get_input_set` . I feel we have a lot of that kind of duplication in the existing pymatgen-custodian-atomate VASP infrastructure and it makes it very difficult to track down where things are being set. I'd prefer you divide them up so that there is only one place to set each thing. We can revisit this decision if it becomes too cumbersome in practice though.

I agree with you that in this context, parameters like temperature and pressure would make sense in `get_input_set`. The generator could define the force field, the integrator, and the fact that this is (for example) an NPT simulation. Specific instances of that recipe can change the Temp or Pressure without fundamentally changing the technical decisions that constitute the recipe for the simulation.

orionarcher replied:
That makes sense to me. In some of my use cases, I'd have to define many OpenMM generators, but I'd say its worth the clarity.

rkingsbury replied:
> 
> 
> That makes sense to me. In some of my use cases, I'd have to define many OpenMM generators, but I'd say its worth the clarity.

I think that's OK, maybe even what we're going for in the generator approach? Since the generators don't store structure, they are lightweight objects that are intended to be customized a lot. The intent is for all the customization to happen there rather than down in `get_input_set`.

I think we'll have to see how this "feels" in practice in order to find the right balance of where to put different settings

orionarcher replied:
Got it. I'll keep this design principle in mind as I finish up the IO.

orionarcher replied:
One thing that *might* belong in `get_input_set` is the `Platform` and `platformProperties` arguments, which are specific to the HPC resource. `Platform` species whether CUDA, OpenCL, or CPU will be used and `platformProperties` can specify properties like GPU device index and number of threads.

Those are a tricky case.

rkingsbury commented on pymatgen/io/openmm/setup.py:
```diff
@@ -0,0 +1,225 @@
+
+class OpenMMSimulationGenerator:
+    """
+    An opinionated generator for OpenMM Simulations sets.
```
What does "opinionated" mean in this context? :)  I'm not familiar.

orionarcher replied:
I suppose I wanted to imply that it does not implement the full OpenMM feature set but instead makes it easier to setup a subset of OpenMM simulations. That language is probably not super clear and redudant with the general Pymatgen approach.

Regardless, I am eventually going to transfer all content in `setup.py` to `inputs.py`. The `setup.py` code was my first pass but is pretty messy and doesn't use the new abstract IO.

rkingsbury commented:
> I believe that the OpenMM IO can fit into the new IO abstraction pretty cleanly by serializing all of the OpenMM objects needed to create a simulation

I appreciate you trying to work within that paradigm; this is a good opportunity to see how it works under a different type of use case. So you're saying that the serialized objects take the place of strings that would be written out to files under a file-based IO paradigm?

> The InputSetGenerator will be instantiated by user defined settings and create a Topology, System, Integrator, and Simulation. It will then serialize all of those objects into XML and store them. When `get_input_set` is called, it will create an `InputSet` that has the XML files as attributes. This process matches the rest of Pymatgen IO and is fit for HTMD.

Does `Topology` depend on the specific system at all? Or does this object provide a generic way of determining topology from structure? If the former, it seems like the `Topology` would need to be created in `get_input_set`.
> 
> The InputSet can then have a `get_simulation` method that instantiates a simulation and returns it. This way, no OpenMM objects are ever stored and the InputSetGenerator and InputSet remain MSONable.

MSONable is good! When you say "instantiates a simulation and returns it", does this actually run the simulation?

orionarcher commented:
> So you're saying that the serialized objects take the place of strings that would be written out to files under a file-based IO paradigm?

Exactly. This lets us save and reconstitute simulations, which as you have pointed out, is quite important for high-throughput applications.

> Does Topology depend on the specific system at all?

Topology is system specific but not `System` specific, if that makes sense. The topology is for a particular set of molecules but not a particular set of coordinates. It only includes bonding information and can be reused if the molecules are the same. The topology has been the trickiest bit to create. Currently, I am using `parmed` to create a structure from SMILES and then generating an `openmm` topology with `parmed`. Perhaps there is a better way that would let us remove the `parmed` dependency.

> If the former, it seems like the Topology would need to be created in get_input_set.

It will be. In fact, *everything* will be created in `get_input_set`. When `get_input_set` is called, InputGenerator will dynamically create the System, Topology, and Integrator, serialize them, and pass them to the InputSet.

> MSONable is good! When you say "instantiates a simulation and returns it", does this actually run the simulation?

It would not! In OpenMM, the simulation does not run until you call `simulation.step(number_of_steps`).

rkingsbury commented:
> > MSONable is good! When you say "instantiates a simulation and returns it", does this actually run the simulation?
> 
> It would not! In OpenMM, the simulation does not run until you call `simulation.step(number_of_steps`).

I see. So maybe the `InputSet` could also have a `run(step=100)` method that would wrap `get_simulation` and actually run it

rkingsbury commented on pymatgen/io/openmm/inputs.py:
```diff
@@ -0,0 +1,301 @@
+        state_file: str = "state.xml",
+    ):
+        topology = TopologyInput.from_file(topology_file)
+        system = TopologyInput.from_file(system_file)
```
should this be `SystemInput.from_file` ?

rkingsbury commented on pymatgen/io/openmm/inputs.py:
```diff
@@ -0,0 +1,301 @@
+    ):
+        topology = TopologyInput.from_file(topology_file)
+        system = TopologyInput.from_file(system_file)
+        integrator = TopologyInput.from_file(integrator_file)
```
should this be `IntegratorInput.from_file`?

rkingsbury commented on pymatgen/io/openmm/inputs.py:
```diff
@@ -0,0 +1,301 @@
+            integrator=integrator,
+        )
+        if Path("state.xml").is_file():
+            state = TopologyInput.from_file(state_file)
```
'StateInput.from_file()`

orionarcher replied:
Thank you. I fixed this and the other typos you identified. Resolving them all.

rkingsbury commented on pymatgen/io/openmm/inputs.py:
```diff
@@ -0,0 +1,301 @@
+
+class OpenMMSet(InputSet):
+
+    # TODO: if there is an optional file should it be missing or should the value be none?
+    @classmethod
+    def from_directory(
```
I assume you're talking about `state.xml`? If I read your current code right, it says "look for a file named `state.xml`, but that file is not required". I think the way you have it is OK. Are you asking whether the `state_file` kwarg should be `None` for an optional file?

orionarcher replied:
You interpreted my comment correctly. I am wondering if setting the `state_file` kwarg to `None` is more appropriate. It would look something like:

```python
    def from_directory(
        cls,
        directory: Union[str, Path],
        topology_file: str = "topology.pdb",
        system_file: str = "system.xml",
        integrator_file: str = "integrator.xml",
        state_file: Optional[str] = None,
    ):
        topology = TopologyInput.from_file(topology_file)
        system = SystemInput.from_file(system_file)
        integrator = IntegratorInput.from_file(integrator_file)
        openmm_set = OpenMMSet(
            topology=topology,
            system=system,
            integrator=integrator,
        )
        if state_file:
            state = StateInput.from_file(state_file)
            openmm_set["state"] = state
        return openmm_set
```

rkingsbury commented on pymatgen/io/openmm/inputs.py:
```diff
@@ -0,0 +1,301 @@
+            openmm_set["state"] = state
+        return openmm_set
+
+    def validate(self) -> bool:
+        # TODO: this should test if the set returns a valid simulation and throw an error if it does not
+        # TODO: is this condition too strict?
```
I don't have direct experience with OpenMM yet, but this sounds right to me. My intention with `validate` is to provide a way to check for errors that would prevent a calc from running properly ( before you have submitted the jobs and waited in a queue). So checking that a valid simulation is returned seems like the right condition to me.

orionarcher replied:
Thanks for the context. We could also call `simulation.step(1)` to propagate a single step, but that would assume that the user can run OpenMM on their local system, which might not be strictly necessary.

rkingsbury commented on pymatgen/io/openmm/inputs.py:
```diff
@@ -0,0 +1,301 @@
+        else:
+            return True
+
+    def get_simulation(self) -> Simulation:
+        simulation = Simulation(
+            self.topology.topology,
+            self.system.system,
+            self.integrator.integrator,
+        )
+        if hasattr(self, "state") and self.state:
+            # TODO: confirm that this works correctly
+            simulation.context.setState(self.state.state)
+        return simulation
```
Per comment above and our discussion about integrating `run` into the IO interface, can you add a `run()` method that calls this and actually executes the simulation?

orionarcher replied:
I definitely could, but I am not sure we should. 

For other codes, like Packmol, `run()` is a simple operation that would be annoying for the user to execute themself. In OpenMM, once the user has a simulation it's very easy for them to run it with `simulation.run(steps)`. Additionally, a user would likely want to do some equilibration, energy minimization, or annealing before running. If we aren't going to support those, I think a `run` method would only cloud the purpose of the InputSet.

To me, it seems cleaner for pymatgen to handle just IO and to leave running the simulation to the user or to `atomate`.

I'm happy to discuss this more though.

rkingsbury replied:
I see. I guess this ties in with our broader discussion of where to put customization settings. For example, we _could_ define all those steps as part of the `InputSet` so that `run()` would still make sense (because all the choices have already been made). But that's not to say we should.

I defer to your judgment; it's fine with me to do without `run` until we see how this works in practice and at scale.

orionarcher replied:
Agreed, and we can certainly defer a final decision until this IO is more final.

> For example, we could define all those steps as part of the InputSet

While this is true for most IO, this is actually not true in the case of OpenMM. Think of the simulation as the starting point. Once we have a simulation, we can use Python to dynamically execute steps, change the temperature, or change the ensemble, but we can't add instructions for those operations in advance. I mean, we could, but we would essentially be dynamically generating a Python script.

That's a key way that OpenMM, and probably other pythonic simulation codes, differ from the traditional approach.

orionarcher committed:
fixed several typos in code

orionarcher committed:
add file for testing_inputs

orionarcher committed:
updated documentation and OpenMMGenerator __init__ and get_input_set

orionarcher committed:
update InputFile approach to remove dynamic OpenMM objects, factor into OpenMMGenerator __init__

orionarcher commented:
A few development updates:

The initial skeleton is complete. Once documentation and testing is written, this will be in a good place.

I've decided to only support the LangevinMiddleIntegrator in the `OpenMMGenerator`. This is a good choice for most simulations. It would be tricky to support other integrators because they all take different kwargs, some of which are not MSONable. If a user wants to use a different integrator, it would be easy to add once they have generated an `InputSet`.

The biggest missing feature is adding partial charges directly from an `.xyz` file and a list of partial charges, which would make it incredibly easy to import RESP charges from DFT. I've discussed this a bit with the OpenFF devs and they have been incredibly helpful, see [this thread](https://github.com/openforcefield/openff-toolkit/issues/1126#issuecomment-967686229).

orionarcher committed:
add testing skeleton and minor syntax changes

orionarcher committed:
update documentation

orionarcher committed:
add testing for OpenMMGenerator

orionarcher committed:
add testing for static methods

orionarcher committed:
add force field file

orionarcher committed:
adding and debugging testing

orionarcher committed:
fixed a bug in the packmol IO

orionarcher committed:
completed first set of testing for OpenMMGenerator

orionarcher committed:
added abstract XmlInput class to simplify InputFile definitions

orionarcher committed:
rewrote input set to better fit new IO abstraction

orionarcher committed:
change naming conventions for InputFiles and attempt to finish get_simulation()

orionarcher committed:
added several test files and datafiles.py to organize test files

orionarcher committed:
add corrected topology.pdg

orionarcher committed:
fixed pdb file bug and removed need for coordinates by defaulting to zeros

orionarcher committed:
finish bug fix for TopologyInput

orionarcher committed:
clean up

orionarcher committed:
more clean up

orionarcher committed:
fix issue with input file syntax using XmlInput

orionarcher committed:
complete testing for InputFiles

orionarcher committed:
added InputSet testing and completed bug fixes. Testing is now complete

orionarcher committed:
another packmol bugfix

orionarcher committed:
improved documentation

orionarcher committed:
finished first pass documentation

orionarcher commented:
I've finished writing testing and documentation for the OpenMM IO. I feel that things are now in a pretty good place and I plan to start using this for my production simulations. I'd love to hear feedback on what could be improved / revised.

There are a few more features and tasks I have in mind:

1. better support for water models, at least allow the user to specify TIP3 or TIP4. I know this is important to @rkingsbury's research, so I'd love to discuss this more.

2. automatic partial charge assignment. It should be possible to include RESP partial charges in the parameterization. I would like to allow either a `.xyz` + partial charge list or a pymatgen.Molecule.

3. good documentation is a journey, not a destination. I'll continue trying to improve the documentation as I develop.

4. (maybe) add support for AMBER, CHARMM, and GAFF force fields. This is something I would need to dig into and not a priority right now.

orionarcher committed:
remove unused setup.py

orionarcher committed:
delete testing for setup.py

orionarcher committed:
added files for testing partial charge assignment

orionarcher committed:
added skeleton for partial charge assignment and testing

orionarcher committed:
rename a directory and propagate changes

orionarcher committed:
restructure datafiles.py and test_inputs.py to remove all hard paths

orionarcher committed:
finish restructuring datafiles and begin testing automated partial charge assignment

orionarcher committed:
added several files to test chirality matching with partial charge assignment

orionarcher committed:
deleted several useless files that I accidentally committed

orionarcher committed:
a initial implementation of partial charge matching

orionarcher committed:
rewrote get_atom_map to progressively relax stererochemistry and bond order restrictions

orionarcher committed:
added testing for infer_openff_mol and get_atom_map

orionarcher committed:
add testing for _assign_charges_to_openff_mol

orionarcher committed:
finished testing for partial charge inferrer

orionarcher committed:
add add_partial_charges_to_forcefield, combining other partial charge handlers into single function to be called by get_input_set

orionarcher committed:
remove equal length assertion in get_atom_map and rename get_box

orionarcher committed:
finish all testing

orionarcher committed:
a couple small bug fixes

orionarcher committed:
a monstrous test that is trying to track down an enigmatic bug, will be removed

orionarcher commented:
Automatic RESP charge assignment is nearly complete. It might even be complete already! All unit tests are passing but my final regression test is failing, however, the regression test might be ill-posed? Currently investigating.

Once automatic RESP charge assignment is complete, it will be possible to add partial charges like this:
```python
CCO_xyz = "geometry_optimizations/CCO.xyz"
CCO_resp_charges = [-0.2894,  0.0095, -0.5882,  0.0915,  0.0915,  0.0915,  0.0939,  0.0939,  0.4058]
generator = OpenMMGenerator(
    partial_charges=[(CCO_xyz, CCO_resp_charges)]
)
input_set = generator.get_input_set({"O": 200, "CCO": 20}, density = 1)
```
The OpenMMGenerator will automatically determine that the molecule in `CCO.xyz` is isomorphic to the "CCO" smile, determine a mapping between them, and assign the partial charges in `CCO_resp_charges` to all "CCO" molecules in the system.

I'm happy with how this turned out. Once this is working, it will be dead easy to add in partial charge information.

EDIT: partial charge matching is now fixed and everything seems to be working well! Once water model support is complete, this will be ready to merge.

orionarcher commented on pymatgen/io/openmm/inputs.py:
```diff
@@ -0,0 +1,622 @@
+            charged_mol = pymatgen.core.Molecule.from_file(str(charged_mol))
+        with tempfile.NamedTemporaryFile() as f:
+            # these next 4 lines are cursed
+            pybel_mol = BabelMolAdaptor(charged_mol).pybel_mol  # pymatgen Molecule
+            pybel_mol.write("mol2", filename=f.name, overwrite=True)  # pybel Molecule
+            rdmol = rdkit.Chem.MolFromMol2File(f.name, removeHs=False)  # rdkit Molecule
+        inferred_mol = openff.toolkit.topology.Molecule.from_rdkit(
+            rdmol, hydrogens_are_explicit=True
+        )  # OpenFF Molecule
```
This function uses a `Molecule` from 4 cheminformatics packages in 4 lines of code. It has to be some sort of record. (commenting because I thought it was funny)

orionarcher committed:
reorganize imports

orionarcher committed:
clean up imports and document imports

orionarcher committed:
finished tracking down a truly heinous bug, atom ordering was reversed in _assign_charges_to_openff_mol, partial charge automation is now complete

orionarcher committed:
removed unused files and correct error in FEC.npy test file

orionarcher committed:
small syntax and formatting changes

rkingsbury committed:
Merge branch 'htmd' into htmd_openmm

rkingsbury commented on pymatgen/io/openmm/tests/test_inputs.py:
```diff
@@ -0,0 +1,313 @@
+
+
+class TestInputFiles:
+    def test_topology_input(self):
+        topology_input = TopologyInput.from_file(topology_path)
+        with tempfile.NamedTemporaryFile() as f:
+            topology_input.write_file(f.name)
+        topology = topology_input.get_topology()
+        assert isinstance(topology, openmm.app.Topology)
+        assert topology.getNumAtoms() == 780
+        assert topology.getNumResidues() == 220
+        assert topology.getNumBonds() == 560
```
For each of the tests in `TestInputFiles`, can you add a line to reconstruct the object from the file you wrote? e.g. `topology = TopologyInput.from_file(f)`

Even though the test begins with `from_file`, it doesn't actually confirm that what is written by `write_file` is identical the file in `test_files`.  Adding this one line and then continuing with the test would confirm this.

orionarcher replied:
Done. Is what I implemented what you had in mind?

rkingsbury replied:
Exactly. Thanks!

rkingsbury commented on pymatgen/io/openmm/inputs.py:
```diff
@@ -0,0 +1,635 @@
+        )
+        if Path(dir / state_file).is_file():
+            openmm_set.inputs[state_file] = StateInput.from_file(dir / state_file)
+            openmm_set.state_file = state_file  # should this be a dict-like assignment?
```
No, this is OK as is. For `InputSet` dict-like assignment will populate the `.inputs` dictionary rather than the attributes. So in the line above, you could also say `openmm_set[state_file] = ...` and have the same effect.

`openmm_set.state_file = ...` should work as expected, as you can set and get arbitrary attributes on an `InputSet`.

orionarcher replied:
Great. Thanks for the clarification.

rkingsbury commented on pymatgen/io/openmm/inputs.py:
```diff
@@ -0,0 +1,635 @@
+        friction_coefficient: int = 1,
+        partial_charges: Optional[
+            List[Tuple[Union[pymatgen.core.Molecule, str, Path], np.ndarray]]
+        ] = None,
```
I think `partial_charges` should go in `get_input_set`, because that will be specific to a particular system, won't it?

orionarcher replied:
As discussed offline, I'm happy either way. If we move `partial_charges` to `get_input_set`, I would advocate also moving `partial_charge_scaling`, `temperature`, and `step_size`. Worth more discussion.

rkingsbury commented on pymatgen/io/openmm/tests/test_inputs.py:
```diff
@@ -0,0 +1,313 @@
+        assert len(state.getPositions(asNumpy=True)) == 780
+
+
+class TestOpenMMSet:
```
Can you please also add a test for `write_input`? I know it's less relevant for OpenMM but would still be good to confirm that it works as expected

rkingsbury commented on pymatgen/io/openmm/inputs.py:
```diff
@@ -0,0 +1,635 @@
+
+
+# noinspection PyMethodOverriding
+class OpenMMGenerator(InputGenerator):
```
In theory, someday there could be multiple OpenMM generators tailored for different simulation needs. As such, I suggest you put `OpenMMGenerator` in a separate `generators.py` file and perhaps rename to something descriptive like `OpenMMMolecularGen` or `OpenMMBaseGen`

orionarcher replied:
I went with OpenMMSolutionGen. The generator I wrote is quite narrow in scope. I hope others will write more!

rkingsbury commented on pymatgen/io/openmm/sets.py:
```diff
@@ -0,0 +1,7 @@
+__version__ = "1.0"
+__maintainer__ = "Orion Cohen"
+__email__ = "orion@lbl.gov"
+__date__ = "Nov 2021"
```
Maybe move `OpenMMSet` into this file, to keep a clean correspondence between the respective files and classes (e.g. one for inputs, one for the `InputSet`, and one for `InputGenerator`)

orionarcher replied:
Great idea! I moved everything into its own files. Much cleaner.

rkingsbury commented on pymatgen/io/openmm/tests/test_simulations.py:
```diff
@@ -0,0 +1,29 @@
+    ethanol_simulation.minimizeEnergy()
+    print(ethanol_simulation.context.getState().getTime())
+    equilibrate_pressure(ethanol_simulation, 300)
+    print(ethanol_simulation.context.getState().getTime())
```
Is there way to actually test that this ran as intended? e.g. `assert ethanol_simulation.context.getState().getTime() == some number` or something like that?

orionarcher replied:
Good suggestion. I added tests for end time and end temperature.

rkingsbury commented on pymatgen/io/openmm/simulations.py:
```diff
@@ -0,0 +1,82 @@
+"""
+This functionality eventually belongs in Atomate, but for the sake of rapidly developing
+a OpenMM IO, I will leave it here for now.
+
+The idea is that each function will operate on a openmm.Simulation and propagate the
+simulation forward in time.
+"""
+
```
The methods look good. What would you think about just attaching these directly to `OpenMMSet`? I don't disagree that similar logic will belong in atomate eventually, but some users (that don't use atomate) might appreciate being able to directly call `anneal` or `equilibrate pressure` from an `InputSet`.

rkingsbury replied:
Could you also add a `minimize_energy` method that just wraps `simulation.minimizeEnergy()`?

orionarcher replied:
We discussed this offline. I am going to leave off minimize_energy for now but I'm happy to revisit this in the future.

rkingsbury commented on pymatgen/io/openmm/simulations.py:
```diff
@@ -0,0 +1,82 @@
+def equilibrate_pressure(
+    simulation: Simulation,
+    steps: int,
+):
```
Add pressure and temperature as kwargs here?

orionarcher replied:
done!

rkingsbury commented on pymatgen/io/openmm/simulations.py:
```diff
@@ -0,0 +1,82 @@
+        system.usesPeriodicBoundaryConditions()
+    ), "system must use periodic boundary conditions for pressure equilibration."
+    barostat_force_index = system.addForce(
+        MonteCarloBarostat(1 * atmosphere, 298 * kelvin, 10)
```
read from kwargs instead of hard-coding

rkingsbury commented on pymatgen/io/openmm/simulations.py:
```diff
@@ -0,0 +1,82 @@
+    Parameters
+    ----------
+    simulation : the openmm.Simulation to propagate.
+    steps : the length of the heating, holding, and cooling stages.
```
Please add units to the docstring

rkingsbury commented on pymatgen/io/openmm/simulations.py:
```diff
@@ -0,0 +1,82 @@
+    simulation : the openmm.Simulation to propagate.
+    temperature : the temperature reached for the holding stage.
+    steps : the length of the heating, holding, and cooling stages.
+    temp_steps : temperature is raised and lowered stepwise in temp_steps pieces.
```
Please add units to the docstring

rkingsbury commented:
Nice work on this @orioncohen ; I'm very excited to start using it. See specific comments above. In addition, please run `black` on all the files and check `mypy` linting errors as well (there may be a few that we can't resolve but mostly this should pass)

A couple of questions:
1. What is the purpose of the .xml file? I guess it adds support for the `smirnoff` force field? I'm a little fuzzy on how many force fields are built in to OpenFF vs. have to be supplied separately; we can discuss offline.
2. Similarly w.r.t the water model specification, let's talk so I can get a better understanding of how and where that would come into play. I'd be fine to merge without that feature if the other stuff is in good shape; we can add it later.

rkingsbury commented:
Also, please pull my latest changes on `htmd` branch

orionarcher committed:
update testing to test from_file reconstruction

orionarcher committed:
removed force field .xml, not currently needed

orionarcher commented:
Thanks for the review! I'll get everything linted and the latest `htmd` changes integrated soon. Do the most recent `htmd` changes include an `openbabel` installation?

> What is the purpose of the .xml file? I guess it adds support for the smirnoff force field? I'm a little fuzzy on how many force fields are built in to OpenFF vs. have to be supplied separately; we can discuss offline.

I was using the .xml file for debugging, it's no longer needed so I just removed it. It may be necessary to add FF xml's to support other FFs. To be honest I am a little fuzzy on it too.

> Similarly w.r.t the water model specification, let's talk so I can get a better understanding of how and where that would come into play. I'd be fine to merge without that feature if the other stuff is in good shape; we can add it later.

Sounds good. The default water model for Sage is TIP3. Sage was parameterized with TIP3 in mind, so it's probably the best option unless our benchmarking demonstrates otherwise. We can leave off more complex support for now.

There are two more things I'd like to do before merging:

1. Add partial charge scaling. I've almost finished this on my local fork. When complete, users will be able to specify the charge scaling of particular input molecules with SMILEs, e.g.

```python
CCO_xyz = "geometry_optimizations/CCO.xyz"
CCO_resp_charges = [-0.2894,  0.0095, -0.5882,  0.0915,  0.0915,  0.0915,  0.0939,  0.0939,  0.4058]
generator = OpenMMGenerator(
    partial_charge_scaling={"O": 0.8, "CCO": 0.9}
    partial_charges=[(CCO_xyz, CCO_resp_charges)]
)
input_set = generator.get_input_set({"O": 200, "CCO": 20}, density = 1)
```

2. Clean up the code. There are a few too many private helper functions flying around. For the sake of maintainability and extensibility, I should cleanup and streamline everything a bit better.

orionarcher committed:
Merge remote-tracking branch 'rkingsbury/htmd' into htmd_openmm

# Conflicts:
#	pymatgen/io/packmol.py

orionarcher committed:
Merge branch 'htmd_openmm' of github.com:orioncohen/pymatgen into htmd_openmm

rkingsbury commented on pymatgen/io/packmol.py:
```diff
@@ -83,7 +83,7 @@ def run(self, path: Union[str, Path], timeout=30):
             raise ValueError("Packmol failed with errorcode {} and stderr: {}".format(e.returncode, e.stderr)) from e
         else:
-            with open(Path(path, self.stdoutfile), "w") as out:
+            with open(self.stdoutfile, "w") as out:
```
I'm not sure what happened here; I think the correct packmol implementation should already be in `htmd` (i.e., the one with `open(Path(path, ...))`. If that's not correct let me know.

orionarcher replied:
Not sure what happened here either. Fixed now. The linter did insignificantly modify `packmol.py` though.

rkingsbury commented:
> Thanks for the review! I'll get everything linted and the latest `htmd` changes integrated soon. Do the most recent `htmd` changes include an `openbabel` installation?

Not specifically. `htmd` handles openbabel the same way upstream pymatgen does.

orionarcher committed:
changed partial charge logic to support partial charge scaling

orionarcher committed:
added testing for partial charge scaling

orionarcher committed:
format all files with black and run through precommit linter

orionarcher commented:
EDIT: I just saw your comment above. I'll fix all the errors that I can and leave the rest.

Any advice for getting past the linter? I've passed all of the syntax checks but the mypy static type checking is killing me. It's raising errors that are impossible to solve or incorrect. For example:

1. > pymatgen/io/openmm/inputs.py:465: error: Slice index must be an integer or None

and line 465 is:
```python
raw_coordinates = coordinates.loc[:, "x":"z"].values
```
Here it is complaining about non-integer indexing, but this is a pandas array! It seems silly to change my code to pass this test.

2. > inputs.py:330: error: Signature of "get_input_set" incompatible with supertype "InputGenerator"

Of course it is! I'm overwriting an abstract method. I don't see how this could pass. This would also produce a similar issue for any code that uses the abstract IO.

3. > inputs.py:357: error: Argument 2 to "get_box" of "OpenMMGenerator" has incompatible type "Optional[float]"; expected "float"

The box variable gets used by a function that takes a `float`, but it is redefined before that usage if `None` is given.

Is there any way to relax the static type checker? It seems to me that these issues are impossible/unecessary to fix.

rkingsbury commented:
> EDIT: I just saw your comment above. I'll fix all the errors that I can and leave the rest.
> 
> Any advice for getting past the linter? I've passed all of the syntax checks but the mypy static type checking is killing me. It's raising errors that are impossible to solve or incorrect. For example:
> 
>     1. > pymatgen/io/openmm/inputs.py:465: error: Slice index must be an integer or None
> 
> 
> and line 465 is:
> 
> ```python
> raw_coordinates = coordinates.loc[:, "x":"z"].values
> ```
> 
> Here it is complaining about non-integer indexing, but this is a pandas array! It seems silly to change my code to pass this test.

There is some discussion about a related issue [here](https://github.com/python/mypy/issues/2410); seems like pandas doesn't handle this in the best way right now. For this one I would just add `# type: ignore` to the line and that will disable this `mypy` check.

>     2. > inputs.py:330: error: Signature of "get_input_set" incompatible with supertype "InputGenerator"
> 
> 
> Of course it is! I'm overwriting an abstract method. I don't see how this could pass. This would also produce a similar issue for any code that uses the abstract IO.
> 
Yes, this one seems to crop up every time I inherit from `InputSet`. Mypy is _technically_ right here, because the Liskov substitution principle says that any child class should have the same call signature as the parent. In the abstract class, there are no required args, but here there is one. Unfortunately we had to go this route to keep the abstract class flexible - there was no way to define a consistent name or type for the required argument, and there doesn't currently seem to be a way to tell the abstract class or mypy to expect at least one arg of uknowntype. 

The following will satisfy mypy, because the extra arguments added in the child class have default values, so you could use the same call signature to instantiate `Thing` and `AbstractThing`.

```
import abc
from typing import Optional, Sequence, Union, List, Type, Any


class AbstractThing(metaclass=abc.ABCMeta):
    
    @abc.abstractmethod
    def function(self, x: str) -> str:
        pass


class Thing(AbstractThing):

    def __init__(self):
        return

    def function(self,
                 x: str, 
                 y: str = 'anything') -> str:
        return x+str(y)
```

In your case, you could satisfy mypy by setting a default value for all args, e.g.

```
def get_input_set(
        self,
        smiles: Dict[str, int] = None,
        density: Optional[float] = None,
        box: Optional[List] = None,
    ) -> InputSet:
```

But that's kinda clunky because then you'd have to catch the case when `smiles` is `None`. So here again I would just use `# type: ignore`. That's what I did for the packmol IO.

>     3. > inputs.py:357: error: Argument 2 to "get_box" of "OpenMMGenerator" has incompatible type "Optional[float]"; expected "float"

> 
> The box variable gets used by a function that takes a `float`, but it is redefined before that usage if `None` is given.
> 

This one should be fixable. mypy is not complaining about the `box` variable but rather the value of `density`. In this code block, I don't think the `assert` statement counts as a guard statement as far as mypy is concerned, 
```
        if not (density is None) ^ (box is None):, "Density OR box must be included, but not both."
        # TODO: write test to ensure coordinates and topology have the same atom ordering
        # create dynamic openmm objects with internal methods
        topology = self._get_openmm_topology(smiles)
        if not box:
            box = self.get_box(smiles, density)
```

I think if you change to something like:
```
 if density is not None and not box:
      box = self.get_box(smiles, density)
```
the error will disappear.

orionarcher committed:
modified inputs to satisfy mypy static type checking

orionarcher committed:
ignore type checking for pkg_resources

orionarcher committed:
add docstrings to all InputFile objects

orionarcher committed:
add docstring to init

orionarcher commented:
Thank you for the detailed advice @rkingsbury. Linting is now passing! I would say this is ready to go.

The rest of testing is failing because several dependencies are uninstalled in the CI environment.

rkingsbury commented:
> Thank you for the detailed advice @rkingsbury. Linting is now passing! I would say this is ready to go.
> 
> The rest of testing is failing because several dependencies are uninstalled in the CI environment.

Thanks @orioncohen . Can you add the new dependencies to `requirements-optional.txt`? That should pull them into the CI so that at least some of the tests will pass. See https://github.com/rkingsbury/pymatgen/blob/75c413e6bbe0e314d984a69f17e77a188d16197e/.github/workflows/test.yml#L45

orionarcher committed:
update requirements-optional file with new dependencies

orionarcher committed:
updated github workflow to include openmm dependencies

orionarcher committed:
update requirements-optional again

orionarcher committed:
editing requirements-optional

orionarcher committed:
more changes to test.yml

orionarcher committed:
fixed error in test.yml

orionarcher committed:
removed openff import from requirements-optional

rkingsbury commented:
> > Thank you for the detailed advice @rkingsbury. Linting is now passing! I would say this is ready to go.
> > The rest of testing is failing because several dependencies are uninstalled in the CI environment.
> 
> Thanks @orioncohen . Can you add the new dependencies to `requirements-optional.txt`? That should pull them into the CI so that at least some of the tests will pass. See
> 
> https://github.com/rkingsbury/pymatgen/blob/75c413e6bbe0e314d984a69f17e77a188d16197e/.github/workflows/test.yml#L45

I worry I may have confused you. Anything that's pip installable should go in `requirements-optional.txt`, in which case no changes to the `.yml` files are needed. If you need to conda install something, that should go in the `.yml` file. I think it will cause problems if you have the same dependency in both places.

orionarcher committed:
more testing workflow changes

orionarcher committed:
please pass

orionarcher committed:
only conda dependencies in test.yml

orionarcher committed:
add stricter versioning of openmm

orionarcher commented:
Thanks for the clarification. That's what I tried initially but I could not get it to work. I experimented with conda-installing openmm and rdkit because I was running into import errors.

Currently, I am pip installing everything except openff-toolkit, which is not on PyPI. I am getting this error:

```
Traceback:
/usr/share/miniconda/envs/test/lib/python3.8/site-packages/openff/toolkit/topology/molecule.py:42: in <module>
    from openmm import LocalCoordinatesSite, unit
E   ImportError: cannot import name 'LocalCoordinatesSite' from 'openmm' (/home/runner/work/pymatgen/pymatgen/pymatgen/io/openmm/__init__.py)

During handling of the above exception, another exception occurred:
/usr/share/miniconda/envs/test/lib/python3.8/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
pymatgen/io/openmm/tests/test_inputs.py:10: in <module>
    import openff.toolkit.topology
/usr/share/miniconda/envs/test/lib/python3.8/site-packages/openff/toolkit/topology/__init__.py:1: in <module>
    from openff.toolkit.topology.molecule import (
/usr/share/miniconda/envs/test/lib/python3.8/site-packages/openff/toolkit/topology/molecule.py:45: in <module>
    from simtk import unit
/usr/share/miniconda/envs/test/lib/python3.8/site-packages/simtk/__init__.py:2: in <module>
    import openmm.unit as unit
E   ModuleNotFoundError: No module named 'openmm.unit'
```

I am trying again with stricter versioning requirements.

orionarcher committed:
change openmm version to 7.6.2

orionarcher committed:
I think I already tried this but ¯\_(ツ)_/¯

rkingsbury reviewed (changes requested):
See some specific comments throughout.

orionarcher committed:
split InputFiles, InputSet, and InputGenerator across different files

orionarcher committed:
refactor testing into multiple files

orionarcher committed:
update simulation imports and testing

orionarcher committed:
modify packmol to fit base branch and satisfy linter

orionarcher committed:
black simulations and add kwargs to equilibrate pressure

orionarcher committed:
added test for write_inputs

orionarcher committed:
fix simulations functions and add real testing

orionarcher committed:
syntax change and removed randomness from testing by setting fixed packmol random seed

orionarcher committed:
change header in imports

rkingsbury commented on pymatgen/io/openmm/simulations.py:
```diff
@@ -0,0 +1,100 @@
+    Parameters
+    ----------
+    simulation : the openmm.Simulation to propagate.
+    steps : the length of the heating, holding, and cooling stages. Steps is number of steps.
```
Can you please change `the length of the heating, holding, and cooling stages. Steps is number of steps.` to `the length of the pressure equilibration. Steps is number of steps.` ?

orionarcher replied:
Done!

rkingsbury commented on pymatgen/io/openmm/simulations.py:
```diff
@@ -0,0 +1,100 @@
+    Parameters
+    ----------
+    simulation : the openmm.Simulation to propagate.
+    temperature : the temperature reached for the holding stage (Kelvin)
```
It looks like here you only specify the heating (holding) temperature, is that right? Is it assumed that the cooling goes back to 298K? Maybe make that explicit in the docstring, or add a kwarg for the starting/ending temperature.

orionarcher replied:
The system cools to whatever temperature it started at. I added to the docstring to make this clear.

orionarcher committed:
small docstring changes

orionarcher committed:
deleted __init__.py in tests

orionarcher committed:
change path naming conventions to pass remote testing

orionarcher committed:
add platform and platformProperties arguments to get_simulation

orionarcher commented on pymatgen/io/openmm/sets.py:
```diff
@@ -0,0 +1,130 @@
+        else:
+            return True
+
+    def get_simulation(
+        self,
+        platform: Optional[openmm.Platform] = None,
+        platformProperties: Dict[str, str] = None,
+    ) -> Simulation:
```
Hey @rkingsbury, I wanted to draw attention to this change and make sure it fits into the new abstraction. 

For context, OpenMM Simulations have a `platform` property that specifies what computational resource to use, e.g. CUDA, OpenCL, or CPU. The `platformProperties` dictionary specifies things like thread count or GPU device index. In theory, `platform and `platformProperties` should not alter the trajectory, though properties like double vs single precision definitely can.

For the most part, OpenMM will determine the optimal `platform` and sensible `platformProperties`. However in certain edge cases, like parallelizing across multiple GPUs on a single node, it is necessary to specify manually.

Since these are platform specific, they don't seem to belong in the input set. Adding them as optional arguments to `get_simulation` was the cleanest solution I could think of.

orionarcher committed:
add new testing files for Li

orionarcher committed:
bug fixes and code restructuring for generators.py

orionarcher committed:
added test for partial charge scaling and setting on Li/PF6 system

orionarcher committed:
small typo and changing args to kwargs to attempt to fix parallelization bug

orionarcher committed:
another addition intended to help fix a pernicious multiprocessing error

orionarcher committed:
fix typing error in get_simulation docs

orionarcher committed:
added statements for debugging and changed platform syntax

orionarcher committed:
revert previous commit

orionarcher commented on pymatgen/io/openmm/generators.py:
```diff
@@ -0,0 +1,422 @@
+            self.friction_coefficient / picoseconds,
+            self.step_size * picoseconds,
+        )
+        context = Context(system, integrator)
+        context.setPositions(coordinates)
+        state = context.getState(getPositions=True)
```
Creating this `Context` caused me a truly nightmarish bug when trying to run on the Perlmutter GPUs. Somehow, instantiating a `Context` creates some state in the program that persists even after `get_input_set` has run. Later on, when I call `input_set.get_simulation`, the `Simulation` internally try's to create a `Context` that conflicts with the previous `Context`, causing an error and crashing the program. This does not occur locally when I use `OpenCL` or `CPU` as the platform, and it does not occur when I run remotely using `CPU`. Absolute nightmare.

Unfortunately, we can't instantiate a `State` without creating a `Context`, so these lines are integral to the current IO implementation.

There is no issue if we create the input files in one python file and then load them in another, it is only an issue if we create and load them in the same file. So thankfully, this shouldn't be too much of an issue for HTMD. 

I'm tempted to just ignore this issue and make sure the error is well documented. It's enough of an edge case that I don't think it warrants restructuring the IO. Curious to hear what you think though.

This was awful to debug, so I'm thankful to have gotten to the bottom of this error.

orionarcher committed:
added del statement to delete context

orionarcher committed:
attempt to force garbage collection to remove Context reference

orionarcher committed:
revert del and garbage collection changes, they did not solve the issue

orionarcher committed:
moved several static methods from OpenMMSolutionGenerator to utils.py

orionarcher committed:
removed several static methods from OpenMMSolutionGen

Commit:
Merge remote-tracking branch 'orioncohen/htmd_openmm' into ae_htmd

orionarcher committed:
added utility functions for molarity, volume ratio, and mass ratio calculations.

orionarcher committed:
fixed typing error

Commit:
Merge remote-tracking branch 'orioncohen/htmd_openmm' into ae_htmd

orionarcher committed:
changed generators to allow smiles with  character

Commit:
Hasty edits to make things work

Commit:
Merge remote-tracking branch 'orioncohen/htmd_openmm' into ae_htmd

orionarcher committed:
added partial_charge_method to OpenMMSolutionGen

orionarcher committed:
refactor code to assign partial charges with molecules, rather than assigning to forcefield

orionarcher committed:
refactor testing for new partial charge method

orionarcher committed:
clean up function signatures

orionarcher committed:
add option to specify geometries of initial molecule positions, format whole file with black

orionarcher committed:
add testing for coordinate specification logic

Commit:
Mixed ForceField Systems

WIP, first effort at allowing multiple forcefields, e.g. spce for water and sage for small molecules

Commit:
Merge remote-tracking branch 'origin/ae_htmd' into ae_htmd

Commit:
proofreading

deleted print statement from debug

Commit:
Fixing tests

Added periodic boundaries to OpenMM System. Assumed box is in angstroms and manually converted to nanometers because not sure how units work

orionarcher committed:
TODO reminder for needed test

orionarcher committed:
fixed bug: test_get_box now fails with incorrect output

orionarcher committed:
small formatting change

orionarcher committed:
Alex changes

orionarcher committed:
added new files for testing and deleted unused files

orionarcher committed:
modified datafiles to include new test files

orionarcher committed:
added support for more file types to generators initial_geometries property

orionarcher committed:
added more robust testing for initial_geometries feature

orionarcher committed:
add a TODO reminder

orionarcher committed:
Merge remote-tracking branch 'rkingsbury/htmd' into htmd_openmm

# Conflicts:
#	requirements-optional.txt

orionarcher committed:
move get_atom_map and infer_openff_mol to utils

orionarcher committed:
move order_molecule_like_smiles from generators to utils

orionarcher committed:
move get_coordinates from generators to utils

orionarcher committed:
move get_topology from generators to utils

orionarcher committed:
move assign_charges_to_mols and add_mol_charges_to_forcefield from generators to utils

orionarcher committed:
move parameterize_system from generators to utils

Commit:
Support for multiple forcefields

Improvements on forcefield specification for mixed systems and specific water models

Commit:
pre-commit run

Implemented changes in comments, ran pre-commit

Commit:
Addressing Comments

Did my best to address all comments

Commit:
Separating functions

Added function to assign_biopolymer_and_water_ff

Commit:
Adding coveralls

Commit:
Trying to get linting

Commit:
Unit tests + tip4p removal

Adding unit tests for:
1. smiles with formal charge
2. Ensuring different water models assign different parameters
3. Testing mixing of small molecule ff with water ff

Also removed tip4p support because it requires the addition of an extra particle. Need to figure that out

Commit:
pytest fixing

Fixed smiles string in test_formal_charge
Be careful about smiles!!

Commit:
black formatting

ran black

orionarcher committed:
Merge pull request #2 from arepstein/ae_htmd

Support for multiple forcefields
